README.md
# d_view

D'View is a package that provides wigdet for response and other basic needs in a concise way.

## Screenshot

<img src="https://github.com/indratrisnar/d_view/raw/master/pic/d_view1.png" alt="dview_1" height="540">
<img src="https://github.com/indratrisnar/d_view/raw/master/pic/d_view2.png" alt="dview_2" height="540">

## Usage

1. Response

```dart
DView.empty(),
DView.error(),
DView.loadingBar(),
DView.loadingCircle(),
```

2. Space

@Deprecated
spaceHeight() inside Column
spaceWidth() inside Row

```dart
DView.spaceHeight(),
DView.spaceWidth(8),
```

Migarte to:

height() inside Column
width() inside Row

with default size: defaultSpace, set this variable in `main` to set global space for height() and width() size.
DView.defaultSpace = 20;

```dart
DView.height(),
DView.width(8),
```

3. Text

```dart
DView.textTitle("D'View"),
DView.textAction(
    () {
        print('click See All 2');
    },
    text: 'SEMUA',
    size: 20,
    iconRight: Icons.navigate_next,
),
```

# Tutorial :

[Watch](https://www.youtube.com/watch?v=ElfN1Fn8v3U)

Support me for more feature & packages
[Donate](https://www.paypal.com/paypalme/indratrisnar)

Check my app : [Visit](https://indratrisnar.github.io/projects.html)

Check My Tutorial & Course : [Watch](https://www.youtube.com/channel/UC0d_xINEvCtlDCpWfBpnYpA)


CHANGELOG.md
## 0.0.5

- new method `height` and `width`
- Deprecated: `spaceHeight` and `spaceWidth`
- set default size for space: `defaultSpace`

## 0.0.4

- update dependency and src

## 0.0.3

Add nothing widget

## 0.0.2

Add tutorial

## 0.0.1

- TODO: Describe initial release.


test/d_view_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:d_view/d_view.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/d_view.dart
library d_view;

import 'package:flutter/material.dart';

/// Widget for helping response and utility in Flutter
class DView {
  /// default size for space\
  /// 16.0\
  ///
  /// Effect:
  /// - height()
  /// - width()
  static double defaultSpace = 16.0;

  ///SimpleAppBar with left title
  static PreferredSizeWidget appBarLeft(String title) {
    return AppBar(
      title: Text(title),
      titleSpacing: 0,
    );
  }

  ///SimpleAppBar with center title
  static PreferredSizeWidget appBarCenter(String title) {
    return AppBar(
      title: Text(title),
      centerTitle: true,
    );
  }

  @Deprecated('Please use DView.height()')

  /// space like sizedbox for height, default 16
  static Widget spaceHeight([double size = 16]) {
    return SizedBox(height: size);
  }

  @Deprecated('Please use DView.width()')

  /// space like sizedbox for width, default 16
  static Widget spaceWidth([double size = 16]) {
    return SizedBox(width: size);
  }

  /// space like sizedbox for height,\
  /// default: `DView.defaultSpace`
  static Widget height([double? size]) {
    double sizeX = size ?? defaultSpace;
    return SizedBox(
      height: sizeX.isNegative ? 0 : sizeX,
    );
  }

  /// space like sizedbox for width,\
  /// default: `DView.defaultSpace`
  static Widget width([double? size]) {
    double sizeX = size ?? defaultSpace;
    return SizedBox(
      width: sizeX.isNegative ? 0 : sizeX,
    );
  }

  /// response for empty data
  static Widget empty([String data = "There's no data"]) {
    return Center(child: Text(data));
  }

  /// response for nothing widget
  /// usually used for checking like visibility off
  static Widget nothing() => const SizedBox();

  /// response for loading circle, default align t center
  static Widget loadingCircle() {
    return const Center(child: CircularProgressIndicator());
  }

  /// response for loading linear, default align t center
  static Widget loadingBar() {
    return const Center(child: LinearProgressIndicator());
  }

  /// response for error
  static Widget error({
    String data = 'Something went wrong',
    TextStyle? style,
  }) {
    return Center(child: Text(data, style: style));
  }

  /// text for title content
  static Widget textTitle(
    String text, {
    double? size,
    Color? color,
    FontWeight? weight,
  }) {
    return Builder(builder: (context) {
      return Text(
        text,
        style: TextStyle(
          fontSize: size ?? 18,
          fontWeight: weight ?? FontWeight.bold,
          color: color ?? Theme.of(context).colorScheme.onBackground,
        ),
      );
    });
  }

  /// text for action
  static Widget textAction(
    void Function() onTap, {
    String text = 'See All',
    TextStyle? customStyle,
    double size = 16,
    FontWeight weight = FontWeight.bold,
    Color color = Colors.black,
    IconData? iconRight,
    Color iconRightColor = Colors.black,
    double? iconRightSize,
    Widget? customIconRight,
  }) {
    return GestureDetector(
      onTap: onTap,
      child: Row(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Text(
            text,
            style: customStyle ??
                TextStyle(
                  fontWeight: weight,
                  fontSize: size,
                  color: color,
                ),
          ),
          if (customIconRight != null)
            customIconRight
          else
            iconRight == null
                ? const SizedBox()
                : Icon(
                    iconRight,
                    color: iconRightColor,
                    size: iconRightSize,
                  ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:d_view/d_view.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(title: const Text('DView')),
        body: SingleChildScrollView(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              // Empty
              title('empty', false),
              DView.empty(),
              DView.nothing(), // box with empty size
              DView.empty("Belum ada data"),
              // Error
              DView.nothing(), // box with empty size
              title('error'),
              DView.error(),
              DView.error(data: "Terjadi kesalahan"),
              // Loading
              title('loading'),
              DView.loadingBar(),
              DView.loadingCircle(),
              // Space
              title('space'),
              Column(
                children: [
                  Text('a' * 9),
                  DView.spaceHeight(),
                  Text('b' * 9),
                  DView.spaceHeight(4),
                  Text('c' * 9),
                ],
              ),
              const Divider(),
              Row(
                children: [
                  Text('d' * 9),
                  DView.spaceWidth(),
                  Text('e' * 9),
                  DView.spaceWidth(8),
                  Text('f' * 9),
                ],
              ),
              const Divider(),
              Column(
                children: [
                  Container(height: 30, width: 60, color: Colors.blue),
                  DView.spaceHeight(),
                  Container(height: 30, width: 60, color: Colors.blue),
                  DView.spaceHeight(4),
                  Container(height: 30, width: 60, color: Colors.blue),
                ],
              ),

              const Divider(),
              Row(
                children: [
                  Container(height: 30, width: 60, color: Colors.blue),
                  DView.spaceWidth(),
                  Container(height: 30, width: 60, color: Colors.blue),
                  DView.spaceWidth(50),
                  Container(height: 30, width: 60, color: Colors.blue),
                ],
              ),
              title('textTitle'),
              DView.textTitle("Text Title"),
              title('textAction'),
              DView.textAction(
                () {
                  print('click See All 1');
                },
                color: Colors.amber,
              ),
              DView.textAction(
                () {
                  print('click See All 2');
                },
                text: 'SEMUA',
                size: 20,
                iconRight: Icons.navigate_next,
              ),
              Padding(
                padding: const EdgeInsets.all(16),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    DView.textTitle("Popular", color: Colors.black, size: 22),
                    DView.textAction(
                      () {
                        print('More click');
                      },
                      size: 18,
                      text: 'More',
                      iconRight: Icons.navigate_next,
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget title(String title, [bool space = true]) {
    return Column(
      children: [
        if (space) const SizedBox(height: 16),
        ListTile(
          title: Text(title),
          tileColor: Colors.grey[300],
          trailing: const Icon(Icons.arrow_downward),
        ),
      ],
    );
  }
}
